commit 3eac1bcae9284fb8b18f4b82156c0e85ddb004e5
Author: Lennart Poettering <lennart@poettering.net>
Date:   Wed Oct 17 18:37:48 2018 +0200

    core: enforce a limit on STATUS= texts recvd from services

    Let's better be safe than sorry, and put a limit on what we receive.

diff --git a/src/core/service.c b/src/core/service.c
index d203e5ad1..fd10c1875 100644
--- a/src/core/service.c
+++ b/src/core/service.c
@@ -2897,8 +2897,12 @@ static void service_notify_message(Unit *u, pid_t pid, char **tags, FDSet *fds)
                 _cleanup_free_ char *t = NULL;

                 if (!isempty(e)) {
-                        if (!utf8_is_valid(e))
-                                log_unit_warning(u, "Status message in notification message is not UTF-8 clean.");
+                        /* Note that this size limit check is mostly paranoia: since the datagram size we are willing
+                         * to process is already limited to NOTIFY_BUFFER_MAX, this limit here should never be hit. */
+                        if (strlen(e) > STATUS_TEXT_MAX)
+                                log_unit_warning(u, "Status message overly long (%zu > %u), ignoring.", strlen(e), STATUS_TEXT_MAX);
+                        else if (!utf8_is_valid(e))
+                                log_unit_warning(u, "Status message in notification message is not UTF-8 clean, ignoring.");
                         else {
                                 t = strdup(e);
                                 if (!t)
diff --git a/src/core/service.h b/src/core/service.h
index 7da0a9396..b16494c22 100644
--- a/src/core/service.h
+++ b/src/core/service.h
@@ -238,3 +238,5 @@ NotifyState notify_state_from_string(const char *s) _pure_;

 const char* service_result_to_string(ServiceResult i) _const_;
 ServiceResult service_result_from_string(const char *s) _pure_;
+
+#define STATUS_TEXT_MAX (16U*1024U)
